php/class-widget-posts.php

```diff
@@ -0,0 +1,770 @@
+	 *
+	 * @param string $id_base
+	 * @param array $new_instance
+	 * @param array $old_instance
+	 * @return array
+	 *
+	 * @throws Exception
+	 */
+	public function sanitize_instance( $id_base, $new_instance, $old_instance = array() ) {
+		if ( ! array_key_exists( $id_base, $this->widget_objs ) ) {
+			throw new Exception( "Unrecognized widget id_base: $id_base" );
+		}
+		if ( ! is_array( $new_instance ) ) {
+			throw new Exception( 'new_instance data must be an array' );
+		}
+		if ( ! is_array( $old_instance ) ) {
+			throw new Exception( 'old_instance data must be an array' );
+		}
+		$widget_obj = $this->widget_objs[ $id_base ];
+		// @codingStandardsIgnoreStart
+		$instance = @ $widget_obj->update( $new_instance, $old_instance ); // silencing errors because we can get undefined index notices
+		// @codingStandardsIgnoreEnd
+		return $instance;
+	}
+
+	/**
+	 * Create a new widget instance.
+	 *
+	 * @param string $id_base
+	 * @param array $instance
+	 * @return \WP_Post
+	 *
+	 * @throws Exception
+	 */
+	function insert_widget( $id_base, $instance = array() ) {
+		if ( ! array_key_exists( $id_base, $this->widget_objs ) ) {
+			throw new Exception( "Unrecognized widget id_base: $id_base" );
+		}
+
+		$instance = $this->sanitize_instance( $id_base, $instance );
+		$widget_number = $this->plugin->widget_number_incrementing->incr_widget_number( $id_base );
+		$post_arr = array(
+			'post_name' => "$id_base-$widget_number",
+			'post_content' => wp_json_encode( $instance, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES ), // For search indexing and post revision UI.
+			'post_content_filtered' => static::encode_post_content_filtered( $instance ),
+			'post_status' => 'publish',
+			'post_type' => static::INSTANCE_POST_TYPE,
+		);
+		$r = wp_insert_post( $post_arr, true );
+		if ( is_wp_error( $r ) ) {
+			throw new Exception( sprintf(
+				'Failed to insert/update widget instance "%s": %s',
+				$post_arr['post_name'],
+				$r->get_error_code()
+			) );
+		}
+		$post = get_post( $r );
+		return $post;
+	}
+
+	/**
+	 * Update an existing widget.
+	 *
+	 * @param string $widget_id
+	 * @param array $instance
+	 * @throws Exception
+	 * @return \WP_Post
+	 */
+	function update_widget( $widget_id, $instance = array() ) {
+		$parsed_widget_id = $this->plugin->parse_widget_id( $widget_id );
+		if ( empty( $parsed_widget_id ) ) {
+			throw new Exception( "Invalid widget_id: $widget_id" );
+		}
+		if ( ! $parsed_widget_id['widget_number'] || $parsed_widget_id['widget_number'] < 2 ) {
+			throw new Exception( "Widgets must start numbering at 2: $widget_id" );
+		}
+
+		$post_id = null;
+		$old_instance = array();
+		$post = $this->get_widget_post( $widget_id );
+		if ( $post ) {
+			$post_id = $post->ID;
+			try {
+				$old_instance = $this->get_widget_instance_data( $post );
+			} catch ( Exception $e ) {
+				$old_instance = array();
+			}
+		}
+		$instance = $this->sanitize_instance( $parsed_widget_id['id_base'], $instance, $old_instance );
```

Comment:
@kasparsd Great point. So what is happening right now is that the widget's `update` callback is getting called twice when the widget is updated via the admin or Customizer. But otherwise, if this method is called programmatically by some other means, specifically `import_widget_instances()`, then it only is called once. But if we know that the instance data is already sanitized, then we should skip calling it again.

So I propose that we add a 3rd `$options` argument to this `update_widge()t` and `insert_widget()` methods, like:

```
$options = wp_parse_args( $options, array(
    'needs_sanitization' => true,
) );
```

And in the places we call `update_widget()` or `insert_widget()` in our codebase, since in all of our uses it should all be pre-sanitized either by the `update` callback having been called or the data coming straight from the `options` via migration, we can then supply `array( 'needs_sanitization' => false )` which would then prevent this sanitization logic from being called by wrapping it in `if ( $options['needs_sanitization'] )`.